### DISCUSSION TITLE
Big performance hit introduced in 37.139.4

### DISCUSSION THREAD

**Discussion by onigoetz** [renovatebot/renovate#27019] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.139.4

### If you're self-hosting Renovate, select which platform you are using.

GitHub Enterprise Server

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

We update Renovate every month.
When doing the update this month, our 900 repositories went from taking 35 minutes to run to 70 minutes.

It took a while to track down, I'm 100% sure the change was introduced in this diff: https://github.com/renovatebot/renovate/compare/37.139.3...37.139.4

Here is a run with Renovate 37.139.4: (I observed the same behaviour with other versions up to 37.163.0)
![image](https://github.com/renovatebot/renovate/assets/309594/458e02e2-6d36-4f34-b700-af2e4e9c0617)
(truncated at the end because I deployed version 37.139.3 before the end of the hour)

And another run with Renovate 37.139.3:
![image](https://github.com/renovatebot/renovate/assets/309594/8fed2289-0f35-4de9-a96c-1e6eeebb1a4a)

Relevant parts of my setup:
* Each repository runs in a separate Docker container
* All repositories are re-scheduled every hour
* 10 repositories run concurrently until the queue is empty
* I used a shared Redis instance for caching

I don't know how I can propose a repro case; my impression is that it is related to Redis.

My 2 cents on the issue:
* on the 4 PRs included in https://github.com/renovatebot/renovate/releases/tag/37.139.4 there is only one that applies to our setup: https://github.com/renovatebot/renovate/issues/26711
* I have the impression that the serialization doesn't work for Redis and that it either fails when setting the values or fails when deserializing them

### Relevant debug logs

There is not a single error in the logs; the relevant part for me is the number of calls to the cache and the overall number of requests.

<details><summary>Logs 37.139.3</summary>

```
DEBUG: Repository timing splits (milliseconds) (repository=...)
       "splits": {"init": 9168, "extract": 990, "lookup": 22308, "onboarding": 1, "update": 61},
       "total": 32705
DEBUG: Package cache statistics (repository=...)
       "get": {"count": 227, "totalMs": 16334, "avgMs": 72, "medianMs": 41, "maxMs": 780},
       "set": {"count": 20, "totalMs": 1140, "avgMs": 57, "medianMs": 34, "maxMs": 340}
DEBUG: http statistics ...
       "totalRequests": 2336
DEBUG: Package lookup durations (repository=...)
       "docker": {"count": 4, "averageMs": 175, "totalMs": 701, "maximumMs": 212},
       "maven": {"count": 60, "averageMs": 1623, "totalMs": 97398, "maximumMs": 14247},
       "npm": {"count": 42, "averageMs": 1459, "totalMs": 61277, "maximumMs": 9488}
 INFO: Repository finished (repository=sq/sq-dlt-earnings-plugin)
       "cloned": true,
       "durationMs": 32705

```

</details>

<details><summary>Logs 37.139.3</summary>

```
DEBUG: Repository timing splits (milliseconds) (repository=...)
       "splits": {"init": 9007, "extract": 847, "lookup": 40074, "onboarding": 1, "update": 77},
       "total": 50221
DEBUG: Package cache statistics (repository=...)
       "get": {"count": 189, "totalMs": 5904, "avgMs": 31, "medianMs": 12, "maxMs": 410},
       "set": {"count": 117, "totalMs": 2494, "avgMs": 21, "medianMs": 8, "maxMs": 139}
DEBUG: http statistics ...
       "totalRequests": 6288
DEBUG: Package lookup durations (repository=...)
       "docker": {"count": 4, "averageMs": 344, "totalMs": 1376, "maximumMs": 425},
       "npm": {"count": 42, "averageMs": 1003, "totalMs": 42132, "maximumMs": 3028},
       "maven": {"count": 60, "averageMs": 3511, "totalMs": 210635, "maximumMs": 16258}

```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
@zharinov would this cache payload change have meant that previously cached entries are invalidated? In which case we'd expect a temporary hit in performance, but a return to normal soon after once the packages are looked up and cached again. Any log messages we could use to check cache hit/performance?

@nabeelsaabna did you observe any performance decrease in the hosted app after 37.139.4 or later were adopted?

**Reply by onigoetz** [reply]
I have figured out what the issue is and did a PR with one possible solution : https://github.com/renovatebot/renovate/pull/27020

I will test it on my own instance and report

**Reply by zharinov** [reply]
This line was a mistake: https://github.com/renovatebot/renovate/blob/e23b3c138536c496de73d527721a554c6c197d81/lib/util/compress.ts#L12

**Reply by onigoetz** [reply]
Should I remove it as part of the PR ?

**Reply by zharinov** [reply]
Thanks for investigation, please let me figure it out

**Reply by onigoetz** [reply]
Sure, thanks for looking into it

**Reply by zharinov** [reply]
https://github.com/renovatebot/renovate/pull/27021

**Comment by rarkins** [comment, marked as answer]
Should be fixed in https://github.com/renovatebot/renovate/releases/tag/37.168.2

@nabeelsaabna can you update the hosted app at the earliest opportunity?

**Reply by onigoetz** [reply]
Hi, I deployed this version this morning and I can confirm that we are back to 35 minutes to process our 900 repositories.

Thanks a lot for the quick response!

**Comment by rarkins** [comment]
@onigoetz thank you very much for reporting the issue with such good analysis and troubleshooting! It helped us to solve it quickly

**Reply by onigoetz** [reply]
@rarkins my pleasure! I just didn't see myself reporting "Renovate is slower when upgrading from 37.100 to 37.153" 
So I started to investigate, one thing led to another and here we are :)

**Reply by rarkins** [reply]
Can you let us know once you can confirm that the new version has solved the regression?

**Reply by onigoetz** [reply]
Hi, I deployed it this morning and I can confirm that we are back to 35 minutes to process our 900 repositories.

Thanks a lot for the quick response!
